internal/status-tracker/src/status_tracker.rs

```diff
@@ -0,0 +1,169 @@
+        }
+    }
+
+    /// Adds a non-error [`LogItem`] to this status tracker.
+    ///
+    /// Primarily intended for use by [`LogItem::success()`]
+    /// or [`LogItem::informational()`].
+    pub fn add_non_error(&mut self, mut log_item: LogItem) {
+        if let Some(ingredient_uri) = self.ingredient_uris.last() {
+            log_item.ingredient_uri = Some(ingredient_uri.to_string().into());
+        }
+        self.logged_items.push(log_item);
+    }
+
+    /// Adds an error-case [`LogItem`] to this status tracker.
+    ///
+    /// Will return `Err(err)` if configured to stop immediately on errors or
+    /// `Ok(())` if configured to continue on errors. _(See [`ErrorBehavior`].)_
+    ///
+    /// Primarily intended for use by [`LogItem::failure()`].
+    pub fn add_error<E>(&mut self, mut log_item: LogItem, err: E) -> Result<(), E> {
+        if let Some(ingredient_uri) = self.ingredient_uris.last() {
+            log_item.ingredient_uri = Some(ingredient_uri.to_string().into());
+        }
+
+        self.logged_items.push(log_item);
+
+        match self.error_behavior {
+            ErrorBehavior::StopOnFirstError => Err(err),
+            ErrorBehavior::ContinueWhenPossible => Ok(()),
+        }
+    }
+
+    /// Returns the [`LogItem`]s that have error conditions (`err_val` is
+    /// populated).
+    ///
+    /// Removes matching items from the list of log items.
+    pub fn filter_errors(&mut self) -> impl Iterator<Item = &LogItem> {
+        self.logged_items()
+            .iter()
+            .filter(|item| item.err_val.is_some())
+    }
+
+    /// Returns `true` if the validation log contains a specific C2PA status
+    /// code.
+    pub fn has_status(&self, val: &str) -> bool {
+        self.logged_items().iter().any(|vi| {
+            if let Some(vs) = &vi.validation_status {
+                vs == val
+            } else {
+                false
+            }
+        })
+    }
+
+    /// Returns `true` if the validation log contains a specific error.
+    pub fn has_error<E: Debug>(&self, err: E) -> bool {
+        let err_type = format!("{:?}", &err);
+        self.logged_items().iter().any(|vi| {
+            if let Some(e) = &vi.err_val {
+                e == &err_type
+            } else {
+                false
+            }
+        })
+    }
+
+    /// Keeps track of the current ingredient URI, if any.
+    ///
+    /// The current URI may be added to any log items that are created.
+    pub fn push_ingredient_uri<S: Into<String>>(&mut self, uri: S) {
+        self.ingredient_uris.push(uri.into());
+    }
+
+    /// Removes the current ingredient URI, if any.
+    pub fn pop_ingredient_uri(&mut self) -> Option<String> {
+        self.ingredient_uris.pop()
+    }
+
+    /// Returns the [`LogItem`]s that have error conditions (`err_val` is
+    /// populated).
+    ///
+    /// Removes matching items from the list of log items.
+    pub fn take_errors(&mut self) -> Vec<LogItem> {
```

Comment:
How is this used? it seems really odd to have something that just extracts the errors

Reply:
I find 23 hits in the repo, all in test cases. Since it's been part of the public API for a while (https://docs.rs/c2pa-status-tracker/latest/c2pa_status_tracker/struct.DetailedStatusTracker.html#method.take_errors), I don't know who else is using it.